## Q1 ##

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('/Users/chloewang/Downloads/2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')

In [2]:
df.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [8]:
df['order_amount'][df.order_amount > 2000].value_counts()

25725     19
704000    17
51450     16
77175      9
154350     1
102900     1
Name: order_amount, dtype: int64

In [6]:
df['order_amount'].shape

(5000,)

a: Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.

Answer: Through exploring the descriptive statistic of this dataset, I found that the number 3145 is exactly the average order value, therefore the calculation itself doesn't go wrong. The reason that this number is not reasonable is that the way used to evaluate this data -- using the mean value" -- is inappropriate. Applying other functions, I found that there are 5000 rows in total, and there are 63 rows who has an order amount that is extremely larger than normal order amount. These outliers are so large that they strongly affect the mean  of the total order value. To avoid the effect of these outliers and better evaluate this data, median value should be utilized.

b: What metric would you report for this dataset?
    
Answer: Median order value.

c: What is its value?
    
Answer: According to the descriptive statistic of this dataset, the median order value should be $284, which is a reasonable number.

## Q2 ##

a: How many orders were shipped by Speedy Express in total?

Desired Output: just a number: the number of orders that were shipped by Speedy Express

Steps:
1. Inner join table Orders and table Shippers
2. Use the "where" condition to only contain the orders whose shipper name is Speedy Express
3. Use "count(*)" to count the entire amount that satisfy the previous steps

Answer:
There are 54 orders that were shipped by Speedy Express

In [ ]:
SELECT COUNT(*) as OrdersShippedBySE
FROM Orders o JOIN Shippers s ON o.ShipperID = s.ShipperID
WHERE s.ShipperName = "Speedy Express"
#OrdersShippedBySE: 54

b: What is the last name of the employee with the most orders?

Desired Output: The last name of an employee who made the most orders, and the number of orders made by him/her

Steps:
1. Create a new table whose columns are the last name of each employee and the number of orders made by each employee

1.1 Inner join table Orders and table Employees

1.2 Group all orders based on EmployeeID
2. Make the number of orders made by each employee in a descending order, and take the first row as result

Answer:
The employee with last name "Peacock" has the most order

In [ ]:
SELECT Employee_Order.LastName
FROM(
SELECT o.EmployeeID, e.LastName, COUNT(*) AS numOfOrder
FROM Orders o JOIN Employees e ON o.EmployeeID = e.EmployeeID
GROUP BY o.EmployeeID) AS Employee_Order
ORDER BY Employee_Order.numOfOrder DESC
LIMIT 1 
#LastName: Peacock

c: What product was ordered the most by customers in Germany?

Desired Output: a product name

Steps:

1. Join four tables: OrderDetails, Orders, Products, and Customers where his/her country is Germany
2. Group the orders based on ProductID, and for each unique ProductID, count how many orders are related
3. Make the product count column in the new table in a descending order
4. Take the first row as result

Answer:
For customers in Germany, Gorgonzola Telino is the product that they tend to order

In [ ]:
SELECT Product_Count_Ger.ProductName
FROM (
SELECT od.ProductID, p.ProductName, COUNT(*) AS product_count
FROM OrderDetails od 
JOIN Orders o ON od.OrderID = o.OrderID
JOIN Products p ON od.ProductID = p.ProductID
JOIN (
SELECT Orders.OrderID, Customers.Country
FROM Orders
JOIN Customers ON Orders.CustomerID = Customers.CustomerID) AS Cus_Order ON od.OrderID = Cus_Order.OrderID
WHERE Cus_Order.Country = "Germany"
GROUP BY od.ProductID) AS Product_Count_Ger
ORDER BY Product_Count_Ger.product_count DESC
LIMIT 1
#ProductName: Gorgonzola Telino